File: runtimes/node/src/kms.ts
/**
 * HelixKMS provides envelope encryption and decryption utility using AWS KMS.
 * This is used to transparently encrypt highly sensitive fields (e.g., PII)
 * within FlatBuffers or Protobuf payloads before broadcasting them.
 */
export class HelixKMS {
    private client: any; // KMSClient
    private keyId: string;
    private region: string;
    private awsKmsModule: any;

    constructor(region: string, keyId: string) {
        this.region = region;
        this.keyId = keyId;
    }

    private async init() {
        if (!this.client) {
            try {
                this.awsKmsModule = await import("@aws-sdk/client-kms");
                this.client = new this.awsKmsModule.KMSClient({ region: this.region });
            } catch (err) {
                console.warn("HelixKMS: @aws-sdk/client-kms not installed. Encryption/Decryption will fail.");
                throw new Error("Missing peer dependency @aws-sdk/client-kms");
            }
        }
    }

    /**
     * Encrypts a raw Uint8Array payload using the provided KMS Key ID.
     */
    async encryptPayload(plaintext: Uint8Array): Promise<Uint8Array> {
        await this.init();
        const command = new this.awsKmsModule.EncryptCommand({
            KeyId: this.keyId,
            Plaintext: plaintext,
        });
        
        try {
            const response = await this.client.send(command);
            if (!response.CiphertextBlob) {
                throw new Error("KMS Encryption failed: Missing CiphertextBlob");
            }
            return response.CiphertextBlob;
        } catch (error) {
            console.error("Helix KMS Encryption Error:", error);
            throw error;
        }
    }

    /**
     * Decrypts an encrypted Uint8Array payload using AWS KMS.
     */
    async decryptPayload(ciphertext: Uint8Array): Promise<Uint8Array> {
        await this.init();
        const command = new this.awsKmsModule.DecryptCommand({
            CiphertextBlob: ciphertext,
        });

        try {
            const response = await this.client.send(command);
            if (!response.Plaintext) {
                throw new Error("KMS Decryption failed: Missing Plaintext");
            }
            return response.Plaintext;
        } catch (error) {
            console.error("Helix KMS Decryption Error:", error);
            throw error;
        }
    }
}

File: runtimes/python/helix_rt/kms.py
import asyncio

try:
    import boto3
except ImportError:
    boto3 = None

class HelixKMS:
    def __init__(self, key_id: str, region_name: str = "us-east-1"):
        if boto3 is None:
            raise ImportError("boto3 is required for HelixKMS")
        self.key_id = key_id
        self.kms_client = boto3.client('kms', region_name=region_name)

    async def encrypt_payload(self, plaintext: bytes) -> bytes:
        loop = asyncio.get_event_loop()
        
        def _encrypt():
            response = self.kms_client.encrypt(
                KeyId=self.key_id,
                Plaintext=plaintext
            )
            return response['CiphertextBlob']
        
        return await loop.run_in_executor(None, _encrypt)

    async def decrypt_payload(self, ciphertext: bytes) -> bytes:
        loop = asyncio.get_event_loop()
        
        def _decrypt():
            response = self.kms_client.decrypt(
                CiphertextBlob=ciphertext,
                KeyId=self.key_id
            )
            return response['Plaintext']
            
        return await loop.run_in_executor(None, _decrypt)
